source/Irrlicht/CQuantNormalCache.cpp

```diff
@@ -0,0 +1,181 @@
+	const auto negativeMask = normal < core::vectorSIMDf(0.0f);
+
+	auto found = normalCacheFor2_10_10_10Quant.find(mapToBarycentric(core::abs(normal)));
+
+	if (found != normalCacheFor2_10_10_10Quant.end() && (found->first == mapToBarycentric(core::abs(normal))))
+	{
+		const uint32_t absVec = found->second;
+		auto vec = core::vectorSIMDu32(absVec, absVec >> quantizationBits, absVec >> quantizationBits * 2)& xorflag;
+
+		return restoreSign<uint32_t>(vec, xorflag, negativeMask, quantizationBits);
+	}
+
+	core::vectorSIMDf fit = findBestFit(quantizationBits, normal);
+
+	auto absIntFit = core::vectorSIMDu32(core::abs(fit)) & xorflag;
+
+	uint32_t absBestFit = absIntFit[0] | (absIntFit[1] << quantizationBits) | (absIntFit[2] << (quantizationBits * 2u));
+
+	normalCacheFor2_10_10_10Quant.insert(std::make_pair(mapToBarycentric(core::abs(normal)), absBestFit));
+
+	return restoreSign<uint32_t>(absIntFit, xorflag, negativeMask, quantizationBits);
+}
+
+uint32_t CQuantNormalCache::quantizeNormal8_8_8(const core::vectorSIMDf& normal)
+{
+	constexpr uint32_t quantizationBits = 8u;
+	const auto xorflag = core::vectorSIMDu32((0x1u << quantizationBits) - 1u);
+	auto negativeMask = normal < core::vectorSIMDf(0.f);
+
+	auto found = normalCacheFor8_8_8Quant.find(mapToBarycentric(normal));
+	if (found != normalCacheFor8_8_8Quant.end() && (found->first == mapToBarycentric(normal)))
+	{
+		const auto absVec = core::vectorSIMDu32(found->second.x, found->second.y, found->second.z);
+
+		return restoreSign<uint32_t>(absVec, xorflag, negativeMask, quantizationBits);
+	}
+
+	core::vectorSIMDf fit = findBestFit(quantizationBits, normal);
+
+	auto absIntFit = core::vectorSIMDu32(core::abs(fit)) & xorflag;
+
+	Vector8u bestFit = { absIntFit[0], absIntFit[1], absIntFit[2] };
+
+	normalCacheFor8_8_8Quant.insert(std::make_pair(mapToBarycentric(normal), bestFit));
+
+	return restoreSign<uint32_t>(absIntFit, xorflag, negativeMask, quantizationBits);
+}
+
+uint64_t CQuantNormalCache::quantizeNormal16_16_16(const core::vectorSIMDf& normal)
+{
+	constexpr uint32_t quantizationBits = 10u;
+	const auto xorflag = core::vectorSIMDu32((0x1u << quantizationBits) - 1u);
+	auto negativeMask = normal < core::vectorSIMDf(0.f);
+
+	auto found = normalCacheFor16_16_16Quant.find(mapToBarycentric(normal));
+	if (found != normalCacheFor16_16_16Quant.end() && (found->first == mapToBarycentric(normal)))
+	{
+		const auto absVec = core::vectorSIMDu32(found->second.x, found->second.y, found->second.z);
+
+		return restoreSign<uint64_t>(absVec, xorflag, negativeMask, quantizationBits);
+	}
+
+	core::vectorSIMDf fit = findBestFit(quantizationBits, normal);
+	auto absIntFit = core::vectorSIMDu32(core::abs(fit)) & xorflag;
+
+	Vector16u bestFit = { absIntFit[0], absIntFit[1], absIntFit[2] };
+
+	normalCacheFor16_16_16Quant.insert(std::make_pair(mapToBarycentric(normal), bestFit));
+
+	return restoreSign<uint64_t>(absIntFit, xorflag, negativeMask, quantizationBits);
+}
```

Comment:
these functions are practically identical, the only thing that changes is the number of quantization bits, you could deduce them from the template type of an enum or stored value type

Reply:
Since `quantizeNormal16_16_16` returns uint64_t this solution would require different return type as well:

```
template <E_QUANT_NORM_CACHE_TYPE Type, typename ReturnType>
ReturnType quantizeNormal(const core::vectorSIMDf& normal);
```

and usage would look smth like this:

```
quantizeNormal<E_QUANT_NORM_CACHE_TYPE::Q_2_10_10_10, uint32_t>(vec) ;
quantizeNormal<E_QUANT_NORM_CACHE_TYPE::Q_8_8_8, uint32_t>(vec) ;
quantizeNormal<E_QUANT_NORM_CACHE_TYPE::Q_16_16_16, uint64_t>(vec) ;
```

It's error prone, and kinda wierd imo, I know current solution violates DRY rule but personally i would leave it as it is..

Reply:
Try this

```cpp
template<E_QUANT_NORM_CACHE_TYPE CacheType>
struct vector_for_cache;

template<> struct vector_for_cache<Q_8_8_8>
{
   typedef type uint32_t;
};
template<> struct vector_for_cache<Q_2_10_10_10>
{
   typedef type uint32_t;
};
template<> struct vector_for_cache<Q_16_16_16>
{
   typedef type uint32_t;
};

template<E_QUANT_NORM_CACHE_TYPE CacheType>
using vector_for_cache_t = typename vector_for_cache<CacheType>::type;

template<E_QUANT_NORM_CACHE_TYPE CacheType>
vector_for_cache_t<CacheType> quantizeNormal(const core::vectorSIMDf& normal);
```